tests/directory/directory_test.bzl

```diff
@@ -0,0 +1,158 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit tests for the directory rule."""
+
+load("@bazel_skylib//rules/directory:directory.bzl", "directory")
+load("@bazel_skylib//rules/directory:providers.bzl", "DirectoryInfo")
+load("@rules_testing//lib:analysis_test.bzl", "analysis_test", "test_suite")
+load("@rules_testing//lib:truth.bzl", "matching")
+load(":utils.bzl", "directory_subject", "failure_matching", "failure_test")
+
+def _source_root_test_impl(env, targets):
+    f1 = targets.f1.files.to_list()[0]
+    f2 = targets.f2.files.to_list()[0]
+
+    env.expect.that_collection(targets.root.files.to_list()).contains_exactly(
+        [f1, f2],
+    )
+
+    human_readable = str(targets.root.label)
+
+    root = directory_subject(env, targets.root[DirectoryInfo])
+    root.entries().keys().contains_exactly(["testdata"])
+    root.transitive_files().contains_exactly([f1, f2]).in_order()
+    root.human_readable().equals(human_readable)
+    env.expect.that_str(root.actual.path + "/testdata/f1").equals(f1.path)
+
+    testdata = directory_subject(env, root.actual.entries["testdata"])
+    testdata.entries().keys().contains_exactly(["f1", "subdir"])
+    testdata.human_readable().equals(human_readable + "/testdata")
+
+    subdir = directory_subject(env, testdata.actual.entries["subdir"])
+    subdir.entries().contains_exactly({"f2": f2})
+    subdir.transitive_files().contains_exactly([f2])
+    env.expect.that_str(subdir.actual.path + "/f2").equals(f2.path)
+
+def _source_root_test(name):
```

Comment:
Invert the order. First a test function, then impl. This way you can follow arrange, act, assert parts; reading it from top to bottom.